src/layout.rs

```diff
@@ -185,8 +207,100 @@ impl Layout {
                 }
             }
         }
+        self.process_sequences();
         self.keycodes()
     }
+    /// Takes care of draining and populating the `sequenced` ArrayDeque,
+    /// giving us sequences (aka macros) of nearly limitless length!
+    fn process_sequences(&mut self) {
```

Comment:
You can really simplify this. You only need one variant at a time, having a deque of 16 is really overkill for memory, and for complexity of code.

You can do everything with

```
    cur_event: Option<Event>,
   awaiting_events: &'static [Event],
```

And when you need another event, something like

```rust
match self.awaiting_events {
    [e, tail @ ..] => {
        self.cur_event = Some(e);
        self.awaiting_event = tail;
    }
    [] => (),
}
```

Reply:
I'll see if I can make that work...  I actually tried something like this initially but couldn't get past the borrow checker complaining about not knowing the length of the events ahead of time...  If I only had one sequence I could get it to work but if there were multiple sequences of different lengths it would complain because the size of the events array would be different every time.

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=25221a8bfda08ec6eb055ea5a3efcdb7

You can also have a `ArrayDeque<&'static [Event]>` if you need several slots.

Reply:
https://doc.rust-lang.org/stable/std/primitive.slice.html#method.split_first is more appropriate.